File: 3213-count-subarrays-where-max-element-appears-at-least-k-times/count-subarrays-where-max-element-appears-at-least-k-times.java
class Solution {
    public long countSubarrays(int[] nums, int k) {
        int n=nums.length;
        int mx=0;
        for (int i = 0; i < nums.length; i++) 
            if (nums[i] > mx) 
                mx = nums[i];
        int start=0; int end=0; long cnt=0; int MaxCnt=0;
        while(end<n){
            if(nums[end]==mx){
                MaxCnt++;
            }
            while(MaxCnt==k){
                cnt+=n-end;
                if(nums[start]==mx){
                    MaxCnt--;
                    
                }
                start++;
            }
            end++;
        }
        return cnt;
    }
}

File: 3213-count-subarrays-where-max-element-appears-at-least-k-times/count-subarrays-where-max-element-appears-at-least-k-times.cpp
class Solution {
public:
    long long countSubarrays(vector<int>& nums, int k) {
        int max=0,i=0,j=0,countMax=0;
        int n=nums.size();
        max=*max_element(begin(nums),end(nums));
        long long result=0;
        while(j<n)
        {
            if(nums[j]==max)
                countMax++;
            while(countMax>=k)
            {
                result+=n-j;
                if(nums[i]==max)
                    countMax--;
                i++;
            }
            j++;
        }
        return result;
    }
};
